Add semantic tokens

macmv opened the pull request:
Closes #7450.

This PR adds semantic tokens from language servers to the editor.

With semantic tokens on the left, without on the right:
<img width="3839" height="1986" alt="image" src="https://github.com/user-attachments/assets/849c1307-7e8b-44e5-8ebd-057d1dc475f2" />

I find this a very useful feature for the following reasons:
- Highlighting functions from the standard library differently from user-defined functions makes code a lot easier to scan at a glance. My eyes can just skip over `.map()` and not need to worry that it's a user-defined function.
- Making parameters/local variables a different color is quite handy when I'm trying to parse some dense code.
- Differentiating function calls/properties (specifically in rust) makes code a lot easier to read at a glance.

Overall, while it is a quality of life feature, I find code a lot easier to understand at a glance with semantic tokens.

There are still some things left to finish:
- [x] Remove the semantic token range request. I intended to use it for multibuffers, but I think it's simpler (and performs well enough) to just grab all the semantic tokens for each buffer. Semantic token ranges also cannot be diffed, so we'd need to fetch all the semantic tokens if the user modified the multibuffer at all.
- [x] Improve perf. For large files (10,000+ lines), I am starting to notice a bit of delay when moving the cursor around. This was much worse in my first draft, where I made an anchor for every token, but I still think it can be improved. Right now, the editor feels a lot more clunky, just from that slight extra delay from these tokens.
- [ ] Remote editors. I added the protobufs, but I never actually handled the requests on the server side. I'm not sure if the server should respond with its locally cached tokens or send a fresh request to the LSP server.

Edit: Perf got fixed by clamping the buffer ranges in `create_hightlight_endpoints` correctly. I don't notice any change now.

Aside from that, it works nicely and highlights all the tokens I could think of.

In terms of conflicting with tree-sitter, these highlights are applied ontop of tree-sitter highlights. Additionally, the LSP spec documents how this should be handled. Within the initialization options of the semantic tokens provider, there is the following option:
```ts
	/**
	 * Whether the client uses semantic tokens to augment existing
	 * syntax tokens. If set to `true` client side created syntax
	 * tokens and semantic tokens are both used for colorization. If
	 * set to `false` the client only uses the returned semantic tokens
	 * for colorization.
	 *
	 * If the value is `undefined` then the client behavior is not
	 * specified.
	 *
	 * @since 3.17.0
	 */
	augmentsSyntaxTokens?: boolean;
```

I set this to `true`, so language servers won't send us semantic tokens for obvious things like strings or numbers. In particular, here's how rust-analyzer handles that flag: https://github.com/rust-lang/rust-analyzer/blob/4ae99f0150c94f4bdf7192b4447f512ece3546fd/crates/rust-analyzer/src/lsp/to_proto.rs#L734-L752

This is best reviewed commit-by-commit. Each commit builds on the last, and they're somewhat organized so that the project should compile between each commit.

Release Notes:

- Added Semantic Tokens.

macmv committed:
Add semantic tokens to LspStore

macmv committed:
Add semantic token protobufs

macmv committed:
Add an LSP command to request semantic tokens

macmv committed:
Plumb semantic tokens through SemanticsProvider

macmv committed:
Store semantic tokens in the display map, and wire that through to CustomHighlights

macmv committed:
Add SemanticTokenStylizer, which converts tokens to HighligtStyles

macmv committed:
Update semantic tokens whenever an editor is modified

macmv committed:
Implement PositionedSemanticTokens, which converts semantic tokens into styled ranges

macmv committed:
Add semantic tokens to the debug syntax tree

macmv committed:
Insert highlights in the custom highlight endpoints

macmv committed:
Add tests

macmv committed:
Add MultiBufferSnapshot::excerpts_for_range and MultiBufferExcerpt::contains_partial_buffer_range

macmv committed:
Add a file to manage semantic tokens

macmv committed:
Add SyntaxTheme::get_opt

SomeoneToIgnore commented on crates/editor/src/display_map/custom_highlights.rs:
```diff
@@ -108,8 +119,43 @@ fn create_highlight_endpoints(
+            }
+        }
+    }
+    highlight_endpoints.sort();
```
With 

```rs
impl Ord for HighlightEndpoint {
    fn cmp(&self, other: &Self) -> cmp::Ordering {
        self.offset
            .cmp(&other.offset)
            .then_with(|| self.style.is_some().cmp(&other.style.is_some()))
    }
}
```

this part of code starts to become more dangerous with each extra semantics layer added.

The main question here to me is: how to unite both tree-sitter and semantics highlight so that people can customize them?

It is not unrealistic to assume there are people who want 0 tree-sitter used in such highlights: `augmentsSyntaxTokens` was mentioned, but no code is added to the configuration level yet.

Would it make sense to use a similar approach

https://github.com/zed-industries/zed/blob/1e58fe77d280e76ead3a7b0802342e6d5daa10fa/assets/settings/default.json#L1519-L1528

to highlights too?

SomeoneToIgnore commented on crates/editor/src/display_map/custom_highlights.rs:
```diff
@@ -108,8 +119,43 @@ fn create_highlight_endpoints(
+            let buffer_range = excerpt
+                .buffer()
+                .range_to_version(excerpt.map_range_to_buffer(range.clone()), &tokens.version);
+            for token in tokens.tokens_in_range(buffer_range) {
```
> tokens_in_range

I know that certain highlights (e.g. word highlight on hover) are applied after a debounce.
Based on my experience of fixing various obscure bugs like https://github.com/zed-industries/zed/pull/40738 , it's not unreasonable to assume that editing and highlights might interleave in such manner, that there is a buffer's `Global` version mismatch between what's highlighted and what's stored in the data.

Here, it seems relatively harmless as would not panic due to anchor/snapshot manipulations, but still potentially buggy?
If so, some extra version check could help — ideally, accessors like

https://github.com/zed-industries/zed/blob/f0ac54e8a576ee830b64ece1dcd969b4254dca78/crates/project/src/lsp_store.rs#L12261-L12274

could help?

SomeoneToIgnore commented on crates/editor/src/display_map.rs:
```diff
@@ -1514,6 +1540,228 @@ impl ToDisplayPoint for Anchor {
+            })
+            .collect::<Vec<_>>();
+
+        // These should be sorted, but we rely on it for binary searching, so let's be sure.
```
NIT: given `.collect::<Vec<_>>();`, it's also possible to use `binary_search_by` for insertion into a `Vec::new()` as a "single" operation instead of two.

Veykril replied:
Collect and sort is likely cheaper given the sheer amount of tokens we might be working with. bin search + insert will have to move a lot of elements around otherwise

SomeoneToIgnore commented on crates/editor/src/display_map.rs:
```diff
@@ -1514,6 +1540,228 @@ impl ToDisplayPoint for Anchor {
+    ) -> Option<HighlightStyle> {
+        let has_modifier = |modifier| self.has_modifier(modifiers, modifier);
+
+        // See the VSCode docs [1] and the LSP Spec [2]
```
This all seems misplaced: `display_map.rs` deals with coordinate transformations between different layers (inlays, folds, tabs) from text buffer into a rendered editor.

Adding LSP into the mix this way seems odd, when we have `LspStore` and related modules.
Method's parameters do not need anything editor-specific, it seems?

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -20697,6 +20759,9 @@ impl Editor {
                     self.update_visible_edit_prediction(window, cx);
                 }
+                if let Some(edited_buffer) = edited_buffer {
+                    self.update_semantic_tokens(edited_buffer, window, cx);
```
I think we need to stick close to what inlay hints (`refresh_inlay_hints`) and document colors (`refresh_colors_for_visible_range`) are doing in fresh `main` (this branch is a bit outdated).

Currently, this is the only place where semantic tokens are updated and inlays update on scrolls too (which is a separate topic), but besides that, there are other cases when we should update the tokens:

* `project::Event::RefreshInlayHints`
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokens_refreshRequest request — odd that the PR does not support it, any reasons not to?

* when a language server is stopped

* `project::Event::LanguageServerBufferRegistered` 
Usually, language servers do not reply anything meaningful before the buffer is registered, and new servers may appear and autoregister the buffer

* `multi_buffer::Event::ExcerptsAdded`
This means that a multi buffer had gotten another text part, maybe from a new file

* `multi_buffer::Event::ExcerptsExpanded` 
* settings change
* etc.

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -22590,6 +22655,12 @@ pub trait SemanticsProvider {
 
+    fn semantic_tokens(
+        &self,
+        buffer_handle: Entity<Buffer>,
```
Looking at this signature, it seems that
range
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokens_rangeRequest
requests are not supported?
How does this work (cpu-wise, latency-wise) on files like editor/src/editor_tests.rs when those are edited continuously?
I do not believe always querying a token request for the entire file is sustainable what if they do not support deltas?), as language servers can be slow and Zed might load them with such queries.

macmv replied:
I've removed all the range bits from this change.

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -6670,6 +6675,63 @@ impl Editor {
+                .semantic_tokens(buffer.clone(), cx)
+                .map(move |task| task.map(move |tokens| (tokens, buffer_id)));
+
+            self.update_semantic_tokens_task = cx.spawn_in(window, async move |this, cx| {
```
Thank you for using a similar approach as other parts (and the tests, by the way!), yet we also need to take the debounce bit in.

https://github.com/zed-industries/zed/blob/1e58fe77d280e76ead3a7b0802342e6d5daa10fa/crates/editor/src/editor.rs#L15529-L15541

for example, or other editor tasks.

Since it's a visual thing, it makes sense to get this debounce into new LSP settings section for semantic highlights.

SomeoneToIgnore commented on crates/editor/src/editor_tests.rs:
```diff
@@ -16246,6 +16246,231 @@ fn test_highlighted_ranges(cx: &mut TestAppContext) {
 }
 
+#[gpui::test]
+async fn lsp_semantic_tokens_full_capability(cx: &mut TestAppContext) {
```
Superb that we get the test for this, it's very important for large features like these — thank you.

Let's add a test on a multi buffer?
Without the debouncing, I expect the current code to query as server a few time extra, interesting to see if I'm right or wrong.

SomeoneToIgnore commented on crates/project/src/lsp_store/semantic_tokens.rs:
```diff
@@ -0,0 +1,104 @@
+#[derive(Default, Debug, Clone)]
+pub struct SemanticTokens {
+    /// Each value is:
+    /// data[5*i] - deltaLine: token line number, relative to the start of the previous token
```
Feels worth a few unit test around this logic (maybe, on a level above, but somewhere, we do have to test delta and other cases, e.g. the one described in a [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens)?)

SomeoneToIgnore replied:
Thank you for adding tests both on the LSP part the "unit" tests on logic from this module.

Yet, I have failed to find any "unit" tests on applying delta requests on top of the partial one, and the spec seem to describe some example?

That looked somewhat complex to me, and if I'm not confusing things, would be great to add another test.

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -3527,6 +3530,12 @@ struct CodeLensData {
 
+#[derive(Debug, Default)]
+struct SemanticTokensData {
+    result_id: Option<String>,
```
It's a bit odd to see this optional, we do know which server did we send the request to and got the response from.

macmv replied:
Yup, agreed. This was just so I could `#[derive(Default)]`. Now that `LspData` exists, it should be easy to get rid of this.

SomeoneToIgnore reviewed (commented):
Sorry for the delayed review — I was working on inlay hints to provide an example how such range-related LSP things are planned to be cached and used in the system.

I was not too meticulous about the protocol part, as hope @Veykril will check the PR too at some point.

Overall, seems like a step forward compared to https://github.com/zed-industries/zed/pull/27556 as we have delta support and the editor-rendering part is somewhat more elaborate here.

I think it's perfectly fine to ignore the proto part for now (albeit there are `collab/src/editor_tests.rs` and local ssh to test on locally and progress on that).

To resume my comments, I think those steps need to be done:

* consolidate all LSP data caching and code near `LspStore` instead of `editor` to share it — consider range requests
* iron out editor store: when things are updated, range request story part (invalidation is somewhat a pain for inlay hints)
* add refresh support and multi buffer test

I think, we can try and consider inlay hints, chunks and whatever related logic added recently: it's not fully ironed out to be a generic storage for semantic tokens, but the `LspStore` part seems good enough — editor part seems ok to start off from.

Besides that, one huge gap here is the user-facing part of the feature:

* no settings (in default.json file we have a bunch of scattered entries, we need to pick something near lsp and highlights semantically, and add a way to turn things on/off, switch between tree-sitter, etc.)
* no documentation (we document settings mainly)
* seems worthy to add another entry to 

<img width="270" height="322" alt="Image" src="https://github.com/user-attachments/assets/7c0f61ec-1784-4b9b-9c0e-08445f2bf1e2" />

for the feature visibility.

* last but not least, it would be superb to be able to do things like 

```
"editor.semanticTokenColorCustomizations": {
    "enabled": true,
    "rules": {
      "unresolvedReference": {
        "foreground": "#c93f3f",
        "fontStyle": "bold"
      }
    }
  },
```

as VSCode allows.
This snippet turns rust-analyzer's highlights into a quite good and useful diagnostic.


-------------------------

I plan to work on inlay hints (there's an invalidation bug somewhere) and rainbow brackets further, so no active work and thinking on this PR yet.
Happy to discuss things further.

macmv commented on crates/language_tools/src/syntax_tree_view.rs:
```diff
@@ -403,6 +464,48 @@ impl SyntaxTreeView {
+                .semantic_tokens
+                .get(&buffer_id)?;
+
+            // This is a hack: it just grabs the first server's capabilities, which isn't
```
Oh, good idea. This was mostly meant for testing, but yeah I'll polish it up and add a toggle.

macmv commented on crates/editor/src/display_map.rs:
```diff
@@ -1514,6 +1540,228 @@ impl ToDisplayPoint for Anchor {
+        // [2]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokenTypes
+        let choices: &[&str] = match self.token_type(token_type)? {
+            // Types
+            "namespace" => &["namespace", "module", "type"],
```
My idea was actually to make all of this configurable, but I didn't get to that on my first pass. Would that make sense? Or should we keep a hardcoded list of tokens we know about?

I'll move it to constants at the least and put all this near the LSP store as well.

macmv commented on crates/lsp/src/lsp.rs:
```diff
@@ -801,6 +851,22 @@ impl LanguageServer {
                         dynamic_registration: Some(true),
                     }),
+                    semantic_tokens: Some(SemanticTokensClientCapabilities {
+                        dynamic_registration: Some(true),
```
Oh good point, I missed that.

SomeoneToIgnore reviewed (commented):
It took me eons to get here, no rush (albeit now I'll try to keep up with the pace and review further changes timely).

> Or just cut out the new protobufs entirely?

I'm not sure how possible this is, given that we'll have to use `Lsp*` infra?
We have to use `LspQuery` for properly made local requests, that drags some `proto` types one way or another.
I think we could literally copy paste document colors/inlays here: the latter is a bit more complex as involves some ranges handling, and the former is more or less the same modulo the LSP data is different.

I do not mind removing the layer/ignoring it until the very end if you find the way — in the end, we'll have to get a `collab/src/tests/editor_tests.rs` entry for remote functionality testing, which I can help with.

> do you prefer rebases or merges

I think merges are more realistic given that this branch will be here for a while, but no preferences really — I prefer to concentrate on the feature itself and its semantics more.

macmv commented:
Alrighty, that sounds good. I'll polish this up and update this soon.

macmv commented on crates/editor/src/display_map/custom_highlights.rs:
```diff
@@ -108,8 +119,43 @@ fn create_highlight_endpoints(
+                }
+                let token_range = excerpt.map_range_from_buffer(token_range);
+
+                if token_range.end <= range.start || token_range.start >= range.end {
```
yeah good point, I think `<` and `>` are correct

macmv commented:
I've posted a new version. This debounces updates in LspStore and aggregates semantic tokens from multiple servers correctly. This largely copies the behavior of document tokens. Notably, when requesting from multiple language servers, semantic tokens are only shown once all servers reply. For each server, it decides to send a full/delta request accordingly.

I added some more tests:
- One test has a multibuffer that shows the same buffer twice. This test shows that the deduplication in lsp store is definitely needed, and it works as expected.
- One test has a buffer with two language servers attached, and this tests aggregating tokens from both servers. The data structures are in place to make this deterministic (there's an `IndexMap<LanguageServerId, BufferSemanticTokens>`), but right now the server that takes priority is the one that responds slowest (I think?).
- One test has a multibuffer with two different languages and language servers attached to each. This is mostly just combining everything and seeing what happens.

I still need the `update_semantic_tokens` task in the editor so that I can aggregate all the buffer changes for multibuffers (as LspStore is only concerned with individual buffers). This task is not debounced at all, as it relies on the lsp store debouncing.

Additionally, `latest_lsp_data` tripped me up a bit, as it clears the previous lsp data every time the buffer changes. I changed this to keep semantic tokens around when updating the `BufferLspData`, but I'm not sure if that makes sense.

And next, I'm going to work on configs. I haven't started that yet.

reflectronic commented:
Just leaving my two cents: I hope that we are not locking ourselves out of implementing `textDocument/semanticTokens/range` support in the future. In particular, I know that the C# language server does not support the `delta` method at this time (https://github.com/dotnet/roslyn/issues/70536). My understanding is that `delta` performs poorly with large files, because it still requires computing semantic analysis for the entire file on each edit. The `range` request allows the computations to be limited to the visible section of the document (e.g. if semantic analysis is implemented in a lazy manner, like it is in the C# compiler).

SomeoneToIgnore reviewed (commented):
Superb, the new tests and most of the LspStore-related changes look solid, we're definitely on the right track.

One small concern is the `text.rs`-related changes, feels that we can inline this all into highlights-related code?

> Additionally, latest_lsp_data tripped me up a bit, as it clears the previous lsp data every time the buffer changes. I changed this to keep semantic tokens around when updating the BufferLspData, but I'm not sure if that makes sense.

This seems somewhat odd, I've commented around the potentially hacky place to discuss it more.
Just to note, there's `current_lsp_data` to look up what's stored inside without invalidating, so that should help with whatever remaining issues, it seems.

> And next, I'm going to work on configs. I haven't started that yet.

One small note here (given the C#-related language server comment above): maybe, we should disable semantic highlights by default, until we support range requests.
If we add the docs and that editor menu item I've mentioned before, that should be discoverable enough for the time being.

macmv commented on crates/text/src/text.rs:
```diff
@@ -2637,11 +2637,24 @@ impl BufferSnapshot {
         false
     }
 
+    pub fn offset_to_version(&self, offset: usize, version: &clock::Global) -> usize {
```
Yeah sure thing, I'll move it. I think it started out needing private APIs from `BufferSnapshot`, but it doesn't now. I'll move it up to where highlights are built.

macmv committed:
Add a toggle for semantic tokens in the debug tree

macmv committed:
Add a config to toggle semantic tokens

macmv committed:
Update semantic tokens in all the other places

macmv committed:
Handle the semanticTokens/refresh request

macmv committed:
Debounce lsp store update

SomeoneToIgnore commented:
Wow, that's quite the first step for the tokens feature, I'd say.

I wonder if `lsp_semantic_token_customizations` could be `semantic_tokens` under `"global_lsp_settings"` key?
Otherwise, the propsal looks good, same as the `theme` custom key: as an alternative (not insisting) we could use `null` as "nothing, fallback to theme" special value?

I would also double check what VSCode does: we have `vscode_import.rs` module, where it should not be too hard to do the same for the tokens' settings, hopefully.
We do not have to support the whole set of those (would be curious to know the discrepancies, though), just what we have already.

The best part would be to document this both in the json and md docs — I can help with that, if needed, same as the RPC part.
If any tests start to fail suddenly, I can help with that: I think that should be straightforward, except maybe a need to write

```rs
cx.executor().advance_clock(Duration::from_secs(LSP_QUERY_TIMEOUT));
cx.run_until_parked();
```

in places to ensure the tokens are queried for.

Will wait a week or so, and also @Veykril 's review as we're getting close and would be great to see if we've missed anything/confirm the direction.

macmv committed:
Add semantic token rules to the config

macmv committed:
Document semantic tokens

xdBronch commented:
firstly, thank you for working on this, very exciting.
is there a way in the current implementation to have a rule that matches a modifier only? i have things like this in my nvim config that don't seem to be possible here
```lua
vim.api.nvim_command(":hi @lsp.mod.mutable gui=underline")
```
this underlines anything with the `mutable` modifier regardless of the token type but something like this is invalid
```json
  "global_lsp_settings": {
    "semantic_token_rules": [
      {
        "token_modifiers": ["mutable"],
        "underline": "#d4be98ff"
      }
    ]
  },
```
am i missing something? would love to be able to do this. 

the nvim version also inherits the underline color from the text which doesnt seem possible with how the rules are currently done since you give a color rather than a bool.
<img width="242" height="53" alt="image" src="https://github.com/user-attachments/assets/b097810f-01f2-46ac-9953-da5b5078a43c" />

macmv committed:
Allow booleans for underline and strikethrough

macmv committed:
Allow matching multiple rules, and make token_type optional

xdBronch commented:
beautiful, thank you!

SomeoneToIgnore committed:
Merge remote-tracking branch 'origin/main' into semantic-tokens

SomeoneToIgnore committed:
Post-merge fixes

SomeoneToIgnore committed:
Take a style pass

SomeoneToIgnore reviewed (commented):
I'm back and should hopefully concentrate at least 50% of the time to push this to completion.

First, thank you for writing out the `lsp_command.rs` layer nicely, esp. all the version awaits, that helps me a lot when dealing with the remote part.

Bad news here is that I need to rewrite quite some part of the `LspStore` layer: there are version races; we cannot use `server_id`s the way they are used, etc.

Moreover, whatever version + usize offset storage happens around is also something I need to change to `Anchor` and other things, otherwise the panics reported in the comments are inevitable.

Also, we handle refresh requests without reloading the actual tokens?
And never handle the capability registration (whilst saying we support that capability).

There are a few smaller issues, I've marked them all with `TODO kb` and plan to deal one at a time, but welcome to help me with either if needed.

I've also got one extra question about the way we treat the LSP types I'd love to discuss and hope you can help me out with.

--------------------------

Those above are relatively small technicalities which I can deal with one way or another.
One standalone question is whether we should move all the customization settings into the theme settings, away from `defaults.json`.

I think we should, but I have no good ideas about that yet and won't get there until deal with the technical stuff first.

It would be good to investigate how can we append settings this way meanwhile, if you feel like that.

Overall, I am ok with taking over the entire PR if you think it's been enough for you at this point.
If not — checking out the points in any order (esp. the last point here) would help a lot.

SomeoneToIgnore committed:
Merge branch 'main' into semantic-tokens

macmv commented on crates/project/src/lsp_command.rs:
```diff
@@ -4483,6 +4842,78 @@ fn process_full_diagnostics_report(
     }
 }
 
+// Copy of the semantic tokens types, except using `Vec<u32>` instead
```
yes, this is because `lsp-types` has an odd definition. i wasn't aware that you already have a fork, i'm happy to go make changes there instead.

the main reason this change is needed is for deltas. a delta request can look something like "replace range 7..8 in the tokens array with [2]". the `lsp-types` delta response doesn't handle that at all, as it attempts to decode that array `[2]` into a sequence of tokens. so, the only change i've made is applying deltas before chunking into groups of 5. additionally, my change ignores any leftover tokens, instead of failing outright if there is not an even multiple of 5.

macmv replied:
In this section: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens, there is a paragraph describing the intent behind the delta response:

> The delta is now expressed on these number arrays without any form of interpretation what these numbers mean. This is comparable to the text document edits send from the server to the client to modify the content of a file. Those are character based and don’t make any assumption about the meaning of the characters. So [ 2,5,3,0,3, 0,5,4,1,0, 3,2,7,2,0 ] can be transformed into [ 3,5,3,0,3, 0,5,4,1,0, 3,2,7,2,0] using the following edit description: { start: 0, deleteCount: 1, data: [3] } which tells the client to simply replace the first number (e.g. 2) in the array with 3.

The problem is that the `lsp-types` encoding of a `DeltaResponse` does not support that `data: [3]` in the given example. If we just switch everything to `Vec<u32>` instead of `Vec<SemanticToken>`, then deltas are much simpler to apply, and we can handle the given example easily.

macmv replied:
PR to the lsp-types fork: https://github.com/zed-industries/lsp-types/pull/9

SomeoneToIgnore commented:
> so i'm inclined to just not support it?

I think this would be the safest approach here, yes, we hold quite the semantic change already.

Thank you for the heads-up about the anchors: we manage to live with [lots of inlays in memory](https://github.com/macmv/zed/blob/cb4a30c84a9fe010452cd67f1a48e100cf865b72/crates/editor/src/display_map/inlay_map.rs#L26-L36) and do coordinate transformations on every editor action for their and others' anchors — so I think there's a way to add tokens into that 🙂 

I suspect that removing this quadratic-looking bit might help a lot:

https://github.com/macmv/zed/blob/cb4a30c84a9fe010452cd67f1a48e100cf865b72/crates/project/src/lsp_store/semantic_tokens.rs#L98-L106

albeit did not look at anything related in depth yet: I plan to deal with the gaps and server query approach first.

I would appreciate if you consider ways to move all customization part into themes.
Zed has

https://github.com/zed-industries/zed/blob/e2f6422b3e1b4c1901523d9000c136e9d624fa7e/docs/src/themes.md?plain=1#L41-L65

and related machinery to update the theme data in memory as the user changes that.

This is per theme, which seems legit, given how VSCode does this and how colors generally work?

Most of the the default.json changes then transform into to more *.md changes (the comment part + an example on how to override the LSP colors) and into a default settings config somewhere.

macmv commented:
yeah sure thing, i'll work on theming. i'm pretty sure VS code lets you define rules to map semantic tokens, and those rules link to styles within the theme. i'll come up with something that makes sense for zed.

SomeoneToIgnore committed:
Straighten the editor layer

SomeoneToIgnore committed:
Move away tokens' logic into a separate editor module

SomeoneToIgnore commented:
Great, thank you for the attempt.
I propose to wait with this a bit, as the current code seems relatively cursed (e.g. it always runs 1 task per buffer, cancelling other tasks for multi buffers which is not right + all anchors + all other things) and I slowly work on the rewrite.

Alas, I'm being bombarded by some other very important tasks and calls, so this does not move too fast, but I hope to finish it this week.
After we shape the basic code, let's do the perf round on top.

That theme settings bit is quite standalone though, and can be worked on even now.

cartercanedy commented:
> Great, thank you for the attempt. I propose to wait with this a bit, as the current code seems relatively cursed (e.g. it always runs 1 task per buffer, cancelling other tasks for multi buffers which is not right + all anchors + all other things) and I slowly work on the rewrite.

no worries, I'm a bit busy myself, but I've been keeping tabs on this PR for a while, and it seemed like an easy enough win. hopefully it can get retained in your updated diff

SomeoneToIgnore committed:
Use anchors for semantic tokens

SomeoneToIgnore committed:
Merge branch 'main' into semantic-tokens

SomeoneToIgnore committed:
Small fixes

SomeoneToIgnore committed:
Remove the semantic tokens toggle from the editor menu

It's not very clear how to handle language settings there, esp. in the multi buffer case.

SomeoneToIgnore committed:
Invalidate semantic tokens with the rest of the LSP data

SomeoneToIgnore committed:
Move more logic into a semantic tokens submodule

SomeoneToIgnore committed:
Properly pass the server for refresh in the LSP store

SomeoneToIgnore committed:
Invalidate the entries better

SomeoneToIgnore committed:
Properly split into remote and local semantic token fetches

SomeoneToIgnore committed:
Merge branch 'main' into semantic-tokens

SomeoneToIgnore committed:
Finish semantic tokens' remote part

SomeoneToIgnore committed:
Merge branch 'main' into semantic-tokens

Veykril committed:
Fix bad performance in `create_highlight_endpoints`

Veykril committed:
Fix visibility of `update_semantic_tokens_task`

Veykril committed:
Remove unnecesay IndexMap, we do not rely on iteration order

SomeoneToIgnore committed:
Merge branch 'main' into semantic-tokens

Veykril committed:
Merge branch 'main' into semantic-tokens

Veykril committed:
Merge branch 'main' into semantic-tokens

Veykril committed:
Prefer semantic tokens over non semantic highlights
